miniapps/blocksolvers/div_free_solver.hpp

```diff
@@ -0,0 +1,252 @@
+   Array<int> coarsest_ess_hdivdofs;  // coarsest level essential H(div) dofs
+   Array<OperatorPtr> C;              // discrete curl: ND -> RT
+   DFSParameters param;
+};
+
+/// For collecting data needed for the divergence free solver
+class DFSDataCollector
+{
+   RT_FECollection hdiv_fec_;
+   L2_FECollection l2_fec_;
+   ND_FECollection hcurl_fec_;
+   L2_FECollection l2_0_fec_;
+
+   unique_ptr<ParFiniteElementSpace> coarse_hdiv_fes_;
+   unique_ptr<ParFiniteElementSpace> coarse_l2_fes_;
+   unique_ptr<ParFiniteElementSpace> coarse_hcurl_fes_;
+   unique_ptr<ParFiniteElementSpace> l2_0_fes_;
+
+   Array<SparseMatrix> el_l2dof_;
+   const Array<int>& ess_bdr_attr_;
+   Array<int> all_bdr_attr_;
+
+   int level_;
+   int order_;
+   DFSData data_;
+
+   void MakeDofRelationTables(int level);
+   void DataFinalize();
+public:
+   DFSDataCollector(int order, int num_refine, ParMesh *mesh,
+                    const Array<int>& ess_attr, const DFSParameters& param);
+
+   void CollectData();
+
+   const DFSData& GetData() const { return data_; }
+
+   unique_ptr<ParFiniteElementSpace> hdiv_fes_;
+   unique_ptr<ParFiniteElementSpace> l2_fes_;
+   unique_ptr<ParFiniteElementSpace> hcurl_fes_;
+};
+
+/// Abstract solver class for Darcy's flow
+class DarcySolver : public Solver
+{
+protected:
+   Array<int> offsets_;
+public:
+   DarcySolver(int size0, int size1) : Solver(size0 + size1), offsets_(3)
+   { offsets_[0] = 0; offsets_[1] = size0; offsets_[2] = height; }
+   virtual int GetNumIterations() const = 0;
+};
+
+/// Solver for B * B^T
+class BBTSolver : public Solver
+{
+   OperatorPtr BBT_;
+   OperatorPtr BBT_prec_;
+   CGSolver BBT_solver_;
+   bool B_has_nullity_one_;
+public:
+   BBTSolver(const HypreParMatrix &B, bool B_has_nullity_one=false,
+             IterSolveParameters param=IterSolveParameters());
+   virtual void Mult(const Vector &x, Vector &y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Block diagonal solver for A, each block is inverted by direct solver
+class BlockDiagSolver : public Solver
+{
+   mutable SparseMatrix block_dof_;
+   mutable Array<int> local_dofs_;
+   mutable Vector sub_rhs_;
+   mutable Vector sub_sol_;
+   Array<DenseMatrixInverse> block_solver_;
+public:
+   BlockDiagSolver(const OperatorPtr& A, SparseMatrix block_dof);
+   virtual void Mult(const Vector &x, Vector &y) const;
+   virtual void MultTranspose(const Vector &x, Vector &y) const { Mult(x, y); }
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Solver for local problems in SchwarzSmoother
+class LocalSolver : public Solver
+{
+   DenseMatrix local_system_;
+   DenseMatrixInverse local_solver_;
+   const int offset_;
+public:
+   LocalSolver(const DenseMatrix &M, const DenseMatrix &B);
+   virtual void Mult(const Vector &x, Vector &y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// non-overlapping additive Schwarz for saddle point problems
+class SchwarzSmoother : public Solver
```

Comment:
I would rename to `SaddleSchwarzSmoother` or something, this is a more specialized idea than what the name suggests.